### DISCUSSION TITLE
Add Discord Rich Presence

### DISCUSSION THREAD

**Discussion by Kal-xyz** [martpie/museeks#679] [Ideas]
Would love if this was also a feature!

**Comment by martpie** [comment]
I don't know what Discord Rich Presence is, can you elaborate? :)

**Reply by Kal-xyz** [reply]
It's mostly like a "status". It displays what you are currently doing/playing if you are using the Discord Desktop app. Pretty sure they have a whole documentation about it here: https://discord.com/developers/docs/rich-presence/how-to

**Reply by Kal-xyz** [reply]
This is what it looks like when if using Spotify (as an example):
![SpotifyDRP](https://github.com/martpie/museeks/assets/44576050/3bb5875b-748c-490d-97ca-32106f97eac9)

**Reply by martpie** [reply]
This is similar in a way to Last.fm or musicbrainz integrations. It's a bit more complicated than just "let's do it" or "let's not do it". I'll try to summarize my thoughts and findings in a separate common discussion :)

**Comment by dipras** [comment]
Yeah, i think it's great idea
What do you think should be displayed(title, artist, etc)?

**Reply by Kal-xyz** [reply]
I would go for something like this;
Keep the title as the Museeks (Without the DRP part)
The details can be something like **Listening to song** or **Currently playing song**
The state can be the artist or the album, maybe **By artist** or **On album**
The timestamp could either be how much of the song is left or how long it has been going on for.
For the large image I would recommend the album cover and for the small image it could be the Museeks logo or it could just be empty.

Here are some images for reference:
![Image1](https://github.com/martpie/museeks/assets/44576050/4696c808-568e-4e2b-a7b5-54762d04f341)
![Image2](https://github.com/martpie/museeks/assets/44576050/26a0adf3-e1a0-49f5-a236-8acc3e793bfc)

(for the second image I am using a custom app)

**Reply by martpie** [reply]
I'll try to keep things short to offer my perspective on the problem: the biggest problem is not the what/how, the biggest problem is authentication with third-party vendors (here, discord) :)

Museeks needs to redirect you to a discord page that will give generate API secrets & tokens, then redirect to Museeks, and store those securely. 

This also probably requires Museeks to register itself as an app to Discord, which is additional administrative stuff, and another thing to maintain the day Discord deprecates its API for another one (rich presence is deprecated for example).

This is a slippery slope for a music player aiming at keeping things simple.

This does not mean we won't do it, it's just I am currently evaluating the pros and cost (of development, not financial) of building such a feature.

**Reply by Kal-xyz** [reply]
Yeah I understand, my point wasn't like "add this instantly". I know that developing this feature will probably be tricky in some ways, as you mentioned the authentication with third-party vendors and that it also has to generate the secrets and tokens. Either way, I don't really mind if it doesn't get added, just wanted to put it as a suggestion so that it might be implemented sometime in the future. 👍

**Reply by dipras** [reply]
I will try to integrate it with [this](https://github.com/xhayper/discord-rpc) to provide you with information about the pros and cost, so it could be your consideration to make this feature :)

**Reply by Kal-xyz** [reply]
Sure, why not

**Reply by Serial-ATA** [reply]
@martpie If there's still interest in this feature, the [discord_presence](https://crates.io/crates/discord_presence) crate makes it pretty simple to create an rpc client. It's used by [mpd-discord-rpc](https://github.com/JakeStanger/mpd-discord-rpc) which you can check out for an implementation.
